webapp/src/components/onboarding/LLMConfigurationForm.tsx

```diff
@@ -0,0 +1,564 @@
+	const [userSelectedEmbeddingType, setUserSelectedEmbeddingType] = useState(false);
+
+	const handleSetUserSelectedLLMType = () => {
+		setUserSelectedLLMType(true);
+	};
+
+	const handleSetUserSelectedEmbeddingType = () => {
+		setUserSelectedEmbeddingType(true);
+	};
+
+	const fetchDatasourceFormData = async () => {
+		await API.getTeamModels({ resourceSlug }, setTeamModels, null, router);
+	};
+
+	const { control, watch, resetField, handleSubmit, setValue, reset } =
+		useForm<LLMConfigurationFormValues>({
+			defaultValues: {
+				LLMType: modelOptions[0],
+				LLMModel: { label: 'gpt-4o-mini', value: 'gpt-4o-mini' },
+				embeddingModel: { label: 'text-embedding-3-small', value: 'text-embedding-3-small' },
+				embeddingType: modelOptions[0]
+			}
+		});
+
+	const { LLMType, embeddingType, LLMModel, embeddingModel } = watch();
+
+	const modelList =
+		[
+			{ label: null, value: null },
+			...ModelList[LLMType.value]
+				.filter(model => !ModelEmbeddingLength[model])
+				.map(model => ({
+					label: model,
+					value: model,
+					...(model === 'gpt-4o' ? { recommended: true } : {})
+				}))
+		] || [];
+	const embeddingModelList =
+		[
+			{ label: null, value: null },
+			...ModelList[embeddingType?.value]
+				.filter(model => ModelEmbeddingLength[model])
+				.map(model => ({ label: model, value: model }))
+		] || [];
+
+	const isOpenAISelectedLLMType = LLMType.value === ModelType.OPENAI;
+	const isOpenAISelectedEmbeddingType = embeddingType.value === ModelType.OPENAI;
+
+	// const embeddingAPIKeyValue = embeddingAPIKey && embeddingAPIKey.startsWith('sk-') && embeddingAPIKey.length > 10
+	// 	? `${embeddingAPIKey.substring(0, 3)}${'X'.repeat(embeddingAPIKey.length - 3)}`
+	// 	: embeddingAPIKey;
+
+	// const llmAPIKeyValue = llmAPIKey && llmAPIKey.startsWith('sk-') && llmAPIKey.length > 10
+	// 	? `${llmAPIKey.substring(0, 3)}${'X'.repeat(llmAPIKey.length - 3)}`
+	// 	: llmAPIKey;
+
+	const LLMModelRequiredFields = Object.keys(ModelTypeRequirements[LLMType.value])
+		.filter(key => !ModelTypeRequirements[LLMType.value][key].optional)
+		.map(key => {
+			const label = key.replace(/_/g, ' ').replace(/\b\w/g, char => char.toUpperCase());
+			const placeholder = key.endsWith('key') ? 'Paste your API key' : `Enter the ${label}`;
+			return { name: key, label, placeholder };
+		});
+
+	const EmbeddingModelRequiredFields = Object.keys(ModelTypeRequirements[embeddingType?.value])
+		.filter(key => !ModelTypeRequirements[embeddingType?.value][key].optional)
+		.map(key => {
+			const label = key.replace(/_/g, ' ').replace(/\b\w/g, char => char.toUpperCase());
+			const placeholder = key.endsWith('key') ? 'Paste your API key' : `Enter the ${label}`;
+			return { name: `embedding_${key}`, label, placeholder };
+		});
+
+	const { isTablet, isMobile } = useResponsive();
+
+	const onSubmit = async (data: LLMConfigurationFormValues) => {
+		setSubmitting(true);
+
+		if (data.LLMModel.value) {
```

Comment:
Instead of doing (pseudocode):

if(llm) {
modelID = await createModel()
await setDefaultModel({ modelId, 'llm' })
}
if(embedding) {
modelID = await createModel()
await setDefaultModel({ modelId, 'llm' })
}

Instead, put them in a promise.all to run both create+setdefaults in parallel:

await Promise.all([
  async() => {
   //create llm & set default
  },
async() => {
   //create embedding & set default
  }
]);

This will make it significantly faster when saving

Reply:
added a promise list that where promises will be pushed conditionally and Promise.all at the end